portal-bridge/src/census.rs

```diff
@@ -0,0 +1,292 @@
+    // the initialization is just to reach a critical mass of peers so that gossip can begin.
+    async fn init(&mut self) {
+        let random_enr = generate_random_remote_enr().1;
+        let Ok(initial_enrs) = self
+            .subnetwork
+            .recursive_find_nodes(&self.client, random_enr.node_id())
+            .await
+        else {
+            panic!("Failed to initialize network census");
+        };
+
+        for enr in initial_enrs {
+            self.iterate_peer_routing_table(enr.clone()).await;
+        }
+    }
+
+    /// Only processes an enr (iterating through its rfn) if the enr's
+    /// liveness delay has expired
+    async fn process_enr(&mut self, enr: Enr) {
+        // ping for liveliness check
+        if !self.liveness_check(enr.clone()).await {
+            return;
+        }
+        // iterate through the peer's rfn
+        self.iterate_peer_routing_table(enr).await;
+    }
+
+    // iterate peers routing table via rfn over various distances
+    async fn iterate_peer_routing_table(&mut self, enr: Enr) {
+        for distance in 245..257 {
+            let Ok(result) = self
+                .subnetwork
+                .find_nodes(&self.client, enr.clone(), vec![distance])
+                .await
+            else {
+                continue;
+            };
+            for found_enr in result {
+                let _ = self.liveness_check(found_enr).await;
+            }
+        }
+    }
+
+    // Only perform liveness check on enrs if their deadline is up,
+    // since the same enr might appear multiple times between the
+    // routing tables of different peers.
+    async fn liveness_check(&mut self, enr: Enr) -> bool {
+        // check if delay map deadline has expired
+        if let Some(deadline) = self.peers.deadline(&enr.node_id().raw()) {
+            if Instant::now() < deadline {
+                return false;
+            }
+        }
+
+        match self.subnetwork.ping(&self.client, enr.clone()).await {
+            Ok(pong_info) => {
+                self.peers
+                    .insert(enr.node_id().raw(), (enr.clone(), pong_info.data_radius));
+                true
+            }
+            Err(_) => {
+                self.peers.remove(&enr.node_id().raw());
+                false
+            }
+        }
+    }
+
+    // Look up all known interested enrs for a given content id
+    async fn get_interested_enrs(&self, content_id: [u8; 32]) -> Vec<Enr> {
+        self.peers
+            .iter()
+            .filter_map(|(node_id, (enr, data_radius))| {
+                let distance = XorMetric::distance(node_id, &content_id);
+                let data_radius = Distance::from(U256::from(*data_radius));
+                if data_radius >= distance {
+                    Some(enr.clone())
+                } else {
+                    None
+                }
+            })
+            .collect()
+    }
+}
```

Comment:
One of the reasons we want to build the big view model of the bridge is because then we can more efficiently bridge data to the network.

I think we should limit how many Enr's are returned. Because well this model gives us improvements, if there is 1_000_000 enr's we are gossiping data to it will be super slow.